### PR TITLE
Transform mistral workbook definition

### PR DESCRIPTION
**Pull request by m4dcoder**
Allow user to define workbook in native DSL without the st2.action
proxy. Before action execution, st2 transforms the workbook definition
and adds the st2.action proxy to appropriate actions before creating and
updating the workbook in Mistral. A st2-mistral pack with an action to
test the transform is included.

### CODE REVIEW

**Comment by m4dcoder** [comment]
The DSL transform is temporary until ActionProvider is implemented in Mistral. I punt on DSL validation in this iteration. I did have to eval inline action parameters to transform the workbook properly.  To be consistent with the evaluation, I copied two utility functions from Mistral to avoid having to import and depend on the entire mistral module.

File: st2actions/st2actions/runners/mistral/utils.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+LOG = logging.getLogger(__name__)
+
+REGEX_ACTION = re.compile("^[\w\.]+[^=\s\"]*")
+REGEX_ACTION_PARAMS = re.compile("([\w]+)=(\"[^\"]*\"\s*|'[^']*'\s*|"
+                                 "\{[^}]*\}\s*|\[.*\]\s*|[\.,:\w\d\.]+)")
+
+
+def _parse_cmd_and_input(cmd_str):
+    cmd_matcher = REGEX_ACTION.search(cmd_str)
```
Comment (manasdk) [inline st2actions/st2actions/runners/mistral/utils.py]: I am a complete regex n00b so I have no clue what this is doing. Can you add some docs or comments in the method to describe.

File: st2actions/st2actions/runners/mistral/utils.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+def _transform_action(spec, action_key, input_key):
+
+    if action_key not in spec or spec.get(action_key) == 'st2.action':
+        return
+
+    if spec.get(action_key) == 'st2.callback':
+        raise Exception('st2.callback is deprecated.')
+
+    _eval_inline_params(spec, action_key, input_key)
```
Comment (manasdk) [inline st2actions/st2actions/runners/mistral/utils.py]: Please add a comment describing the need for this evaluation.

File: st2actions/st2actions/runners/mistral/utils.py (hunk 1 above)
Comment (m4dcoder) [inline st2actions/st2actions/runners/mistral/utils.py]: I copied this from Mistral @ https://github.com/stackforge/mistral/blob/master/mistral/workbook/base.py#L23.  It's used to evaluate inline action parameters (i.e. action: core.local cmd={$.cmd}).  The regex could have been written better.  But I would fix it at the Mistral side.  I'm just trying to keep the evaluation consistent here.

Comment (manasdk) [inline st2actions/st2actions/runners/mistral/utils.py]: K.

**Comment by manasdk** [comment]
I get the idea that a transform is required although I don't understand the transform. Perhaps it is not worth spending time on the details since you guys expect to deprecate the transformation anyway once the ActionProvider is put in place?

File: contrib/sandbox/packages/st2-mistral/actions/transformer.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import os
+import six
+import yaml
+
+from st2actions.runners import pythonrunner
+from st2actions.runners.mistral import utils
+from st2client import client
+
+
+class MistralDSLTransformer(pythonrunner.Action):
```
Comment (manasdk) [inline contrib/sandbox/packages/st2-mistral/actions/transformer.py]: Why does the impl here and in  st2actions/st2actions/runners/mistral/utils.py look so similar? Perhaps you can call into the utils from here?

Also, what is the purpose of this action? Could the actions have been a separate PR or at least a separate commit? 

Btw, we typically put sandbox action in st2incubator now.

**Comment by lakshmi-kannan** [comment]
Yeah as long as this transform works, I am willing to assume the details as a blackbox. There are tests. So we are good. That damn regex made it to our code :).

**Comment by manasdk** [comment]
Hmm ... perhaps it is better if @lakshmi-kannan or @dzimine take a look at this PR.

File: contrib/sandbox/packages/st2-mistral/actions/transformer.py (hunk 3 above)
Comment (lakshmi-kannan) [inline contrib/sandbox/packages/st2-mistral/actions/transformer.py]: I mean that's what he does. 

from st2actions.runners.mistral import utils

But I agree this could have been a separate PR. I was very confused too.

**Comment by lakshmi-kannan** [comment]
I think the PR could have been split in the following way:

1) Transformer utility and tests
2) Use transformer in mistral runner and change the examples
3) Create a mistral action using that transformer

File: contrib/sandbox/packages/st2-mistral/actions/transformer.py (hunk 3 above)
Comment (m4dcoder) [inline contrib/sandbox/packages/st2-mistral/actions/transformer.py]: I'll move this to the st2-incubator.  I put this here because the mistral pack was placed here first by lakshmi.  I can't use utils.transform_definition() because reading from the Action model is different.  The action here uses the st2client whereas the one in utils uses the st2common persistence directly.

**Comment by m4dcoder** [comment]
Sure it would have been nice with #3.  But it's inconsistent with your early comments before that st2 action should not have access to st2 DB directly.

**Comment by lakshmi-kannan** [comment]
#3 could be create a mistral action.

**Comment by m4dcoder** [comment]
I'm not sure what you mean.  But maybe you can patch afterward.

**Comment by m4dcoder** [comment]
Split to multiple commits and added comments for _eval_inline_params.

File: docs/source/mistral.rst [hunk 4]
```diff
@@ -59,64 +55,33 @@ Let's say we need to upgrade and reboot all the members of a MongoDB replica set
                 - duration
```
Comment (dzimine) [inline docs/source/mistral.rst]: It would be nice to make this worklfow actually working against 'mock' actions (make them part of example),  run this workflow as part of the test, and `literanlinclude` it in the docs.

**Comment by dzimine** [comment]
LGTM.
